src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs

```diff
@@ -13,869 +13,614 @@
+            securityBuffer[1] = new SecurityBuffer(buffer, offset + 16, count - 16, SecurityBufferType.Data);
 
-            protocol = _context.ProtocolName;
+            int errorCode;
+            SecurityBufferType realDataType = SecurityBufferType.Data;
 
-            if (_context.IsServer)
+            if (isConfidential)
             {
-                SecurityContextTokenHandle token = null;
-                try
-                {
-                    token = _context.GetContextToken();
-                    string authtype = _context.ProtocolName;
+                errorCode = SSPIWrapper.DecryptMessage(GlobalSSPI.SSPIAuth, securityContext, securityBuffer, sequenceNumber);
+            }
+            else
+            {
+                realDataType |= SecurityBufferType.ReadOnlyFlag;
+                securityBuffer[1].type = realDataType;
+                errorCode = SSPIWrapper.VerifySignature(GlobalSSPI.SSPIAuth, securityContext, securityBuffer, sequenceNumber);
+            }
 
-                    // TODO #5241: 
-                    // The following call was also specifying WindowsAccountType.Normal, true.
-                    // WindowsIdentity.IsAuthenticated is no longer supported in CoreFX.
-                    result = new WindowsIdentity(token.DangerousGetHandle(), authtype);
-                    return result;
-                }
-                catch (SecurityException)
+            if (errorCode != 0)
+            {
+                if (GlobalLog.IsEnabled)
                 {
-                    // Ignore and construct generic Identity if failed due to security problem.
-                }
-                finally
-                {
-                    if (token != null)
-                    {
-                        token.Dispose();
-                    }
+                    GlobalLog.Print("NTAuthentication#" + LoggingHash.HashString(securityContext) + "::Decrypt() throw Error = " + errorCode.ToString("x", NumberFormatInfo.InvariantInfo));
                 }
+                throw new Win32Exception(errorCode);
             }
 
-            // On the client we don't have access to the remote side identity.
-            result = new GenericIdentity(name, protocol);
-            return result;
+            if (securityBuffer[1].type != realDataType)
+            {
+                throw new InternalException();
+            }
+
+            newOffset = securityBuffer[1].offset;
+            return securityBuffer[1].size;
         }
 
-        internal void CheckThrow(bool authSucessCheck)
+        private static ContextFlagsPal GetContextFlagsPalFromInterop(Interop.SspiCli.ContextFlags win32Flags)
         {
-            if (_exception != null)
+            ContextFlagsPal flags = ContextFlagsPal.Zero;
+            if ((win32Flags & Interop.SspiCli.ContextFlags.AcceptExtendedError) != 0)
             {
-                throw _exception;
+                flags |= ContextFlagsPal.AcceptExtendedError;
```

Comment:
Consider a structure that you can just loop over, rather than repeating this many, many times.
```C#
private static readonly KeyValuePair<Interop.SspiCli.ContextFlags, ContextFlagsPal>[] s_contextFlagMappings =
{
    new KeyValuePair<Interop.SspiCli.ContextFlags, ContextFlagsPal>(Interop.SspiCli.ContextFlags.AcceptExtendedError, ContextFlagsPal.AcceptExtendedError),
    ...
};

...

ContextFlagsPal flags = ContextFlagsPal.Zero;

foreach (KeyValuePair<Interop.SspiCli.ContextFlags, ContextFlagsPal> pair in s_contextFlagMappings)
{
    if ((win32Flags & pair.Key) != 0)
    {
        flags |= pair.Value;
    }
}
```

Perhaps with a using-alias to make the type identifier a wee-bit smaller.